File: rust_exercises/leetcode/Problem738/Solution.java
class Solution {
    public int monotoneIncreasingDigits(int N) {
        List<Integer> digits = new ArrayList<>();
        while (N > 0) {
            digits.add(N % 10);
            N /= 10;
        }
        
        int lo = 0;
        int hi = 1;
        while (hi < digits.size()) {
            if (digits.get(hi) > digits.get(hi - 1)) {
                digits.set(hi, digits.get(hi) - 1);
                for (int i = lo; i < hi; i += 1) {
                    digits.set(i, 9);
                }
                lo = hi;
            }
            hi += 1;
        }
        
        int result = 0;
        for (int i = digits.size() - 1; i >= 0; i -= 1) {
            result = 10 * result + digits.get(i);
        }
        return result;
    }
}

File: rust_exercises/leetcode/Problem738/solution.rs
impl Solution {
    pub fn monotone_increasing_digits(n: i32) -> i32 {
        let mut n = n;
        let mut digits = vec![];
        
        while n > 0 {
            digits.push(n % 10);
            n /= 10;
        }
        
        let mut lo = 0;
        let mut hi = 1;
        
        while hi < digits.len() {
            if digits[hi] > digits[hi - 1] {
                digits[hi] -= 1;
                for i in lo..hi {
                    digits[i] = 9;
                }
                lo = hi;
            }
            hi += 1;
        }
        
        let mut result = 0;
        while let Some(d) = digits.pop() {
            result = result * 10 + d;
        }
        
        result
    }
}
